src/core/global.c

```diff
@@ -303,20 +307,28 @@ static void nn_global_term (void)
 void nn_term (void)
 {
     int i;
+    struct nn_sock *sock;
 
     nn_glock_lock ();
 
     /*  Switch the global state into the zombie state. */
     self.flags |= NN_CTX_FLAG_ZOMBIE;
 
     /*  Mark all open sockets as terminating. */
-    if (self.socks && self.nsocks) {
+    if (self.nsocks != 0) {
         for (i = 0; i != NN_MAX_SOCKETS; ++i)
-            if (self.socks [i])
+            if (self.socks [i] != NULL) {
                 nn_sock_zombify (self.socks [i]);
+            }
     }
 
     nn_glock_unlock ();
+
+    /* Make sure we really close resources, this will cause global
+       resources to be freed too when the last socket is closed. */
+    for (i = 0; i < NN_MAX_SOCKETS; i++) {
+        (void) nn_close_impl (i, 1);
+    }
```

Comment:
Yes this is the right way. We're shutting down so the grab and release is fine here. 

Sent from my iPhone

> On Apr 26, 2016, at 6:10 AM, Jack R. Dunaway <notifications@github.com> wrote:
> 
> In src/core/global.c:
> 
> >      }
> >  
> >      nn_glock_unlock ();
> > +
> > +    /* Make sure we really close resources, this will cause global
> > +       resources to be freed too when the last socket is closed. */
> > +    for (i = 0; i < NN_MAX_SOCKETS; i++) {
> > +        (void) nn_close_impl (i, 1);
> > +    }
> In this loop, each iteration obtains/releases the glock. Is this semantic correct, compared to holding the glock for the entirety of looping through all sockets?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
>